File: treatment/src/main/java/com/hartwig/actin/treatment/trial/TrialConfigModel.java
package com.hartwig.actin.treatment.trial;

import java.io.IOException;
import java.util.List;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;
import com.hartwig.actin.treatment.trial.config.CohortDefinitionConfig;
import com.hartwig.actin.treatment.trial.config.InclusionCriteriaConfig;
import com.hartwig.actin.treatment.trial.config.InclusionCriteriaReferenceConfig;
import com.hartwig.actin.treatment.trial.config.TrialDefinitionConfig;

import org.jetbrains.annotations.NotNull;

public class TrialConfigModel {

    @NotNull
    private final TrialConfigDatabase database;

    @NotNull
    public static TrialConfigModel create(@NotNull String trialConfigDirectory, @NotNull EligibilityFactory eligibilityFactory)
            throws IOException {
        TrialConfigDatabase database = TrialConfigDatabaseReader.read(trialConfigDirectory);
        if (!new TrialConfigDatabaseValidator(eligibilityFactory).isValid(database)) {
            throw new IllegalStateException("Trial config database is not considered valid. Cannot create config model.");
        }
        return new TrialConfigModel(database);
    }

    @VisibleForTesting
    TrialConfigModel(@NotNull final TrialConfigDatabase database) {
        this.database = database;
    }

    @NotNull
    public List<TrialDefinitionConfig> trials() {
        return database.trialDefinitionConfigs();
    }

    @NotNull
    public List<CohortDefinitionConfig> cohortsForTrial(@NotNull String trialId) {
        List<CohortDefinitionConfig> cohorts = Lists.newArrayList();
        for (CohortDefinitionConfig cohort : database.cohortDefinitionConfigs()) {
            if (cohort.trialId().equals(trialId)) {
                cohorts.add(cohort);
            }
        }
        return cohorts;
    }

    @NotNull
    public List<InclusionCriteriaConfig> generalInclusionCriteriaForTrial(@NotNull String trialId) {
        List<InclusionCriteriaConfig> configs = Lists.newArrayList();
        for (InclusionCriteriaConfig config : database.inclusionCriteriaConfigs()) {
            if (config.trialId().equals(trialId) && config.appliesToCohorts().isEmpty()) {
                configs.add(config);
            }
        }
        return configs;
    }

    @NotNull
    public List<InclusionCriteriaConfig> specificInclusionCriteriaForCohort(@NotNull String trialId, @NotNull String cohortId) {
        List<InclusionCriteriaConfig> configs = Lists.newArrayList();
        for (InclusionCriteriaConfig config : database.inclusionCriteriaConfigs()) {
            if (config.trialId().equals(trialId) && config.appliesToCohorts().contains(cohortId)) {
                configs.add(config);
            }
        }
        return configs;
    }

    @NotNull
    public List<InclusionCriteriaReferenceConfig> referencesForTrial(@NotNull String trialId) {
        List<InclusionCriteriaReferenceConfig> configs = Lists.newArrayList();
        for (InclusionCriteriaReferenceConfig config : database.inclusionCriteriaReferenceConfigs()) {
            if (config.trialId().equals(trialId)) {
                configs.add(config);
            }
        }
        return configs;
    }
}

File: trial/src/main/kotlin/com/hartwig/actin/trial/config/TrialConfigModel.kt
package com.hartwig.actin.trial.config

import com.hartwig.actin.datamodel.trial.EligibilityRule
import com.hartwig.actin.trial.interpretation.EligibilityFactory

class TrialConfigModel(
    private val trialDefinitionConfigs: List<TrialDefinitionConfig>,
    private val cohortsByTrialId: Map<String, List<CohortDefinitionConfig>>,
    private val generalInclusionCriteriaByTrial: Map<String, List<InclusionCriteriaConfig>>,
    private val specificInclusionCriteriaByTrialAndCohort: Map<String, Map<String, List<InclusionCriteriaConfig>>>,
    private val referencesByTrialAndId: Map<String, Map<String, InclusionCriteriaReferenceConfig>>,
    val unusedRulesToKeep: Set<EligibilityRule>,
    private val trialDatabaseValidation: TrialConfigDatabaseValidation
) {

    fun validation(): TrialConfigDatabaseValidation {
        return trialDatabaseValidation
    }

    fun trials(): List<TrialDefinitionConfig> {
        return trialDefinitionConfigs
    }

    fun cohorts(): List<CohortDefinitionConfig> {
        return cohortsByTrialId.values.flatten()
    }

    fun cohortsForTrial(trialId: String): List<CohortDefinitionConfig> {
        return cohortsByTrialId[trialId] ?: emptyList()
    }

    fun generalInclusionCriteriaForTrial(trialId: String): List<InclusionCriteriaConfig> {
        return generalInclusionCriteriaByTrial[trialId] ?: emptyList()
    }

    fun specificInclusionCriteriaForCohort(trialId: String, cohortId: String): List<InclusionCriteriaConfig> {
        return specificInclusionCriteriaByTrialAndCohort[trialId]?.get(cohortId) ?: emptyList()
    }

    fun referencesForTrial(trialId: String): Map<String, InclusionCriteriaReferenceConfig> {
        return referencesByTrialAndId[trialId] ?: emptyMap()
    }

    companion object {

        fun create(trialConfigDirectory: String, eligibilityFactory: EligibilityFactory): TrialConfigModel {
            val database = TrialConfigDatabaseReader.read(trialConfigDirectory)
            return createFromDatabase(database, TrialConfigDatabaseValidator(eligibilityFactory))
        }

        fun createFromDatabase(
            database: TrialConfigDatabase, trialConfigDatabaseValidator: TrialConfigDatabaseValidator
        ): TrialConfigModel {
            val specificInclusionCriteriaByTrialAndCohort =
                configsByTrial(database.inclusionCriteriaConfigs.filter { it.appliesToCohorts.isNotEmpty() })
                    .mapValues { inclusionCriteriaConfigsByCohortId(it.value) }

            val referencesByTrialAndId = configsByTrial(database.inclusionCriteriaReferenceConfigs)
                .mapValues { it.value.associateBy(InclusionCriteriaReferenceConfig::referenceId) }

            val validRules = EligibilityRule.values().map(EligibilityRule::toString).toSet()

            return TrialConfigModel(
                database.trialDefinitionConfigs,
                configsByTrial(database.cohortDefinitionConfigs),
                configsByTrial(database.inclusionCriteriaConfigs.filter { it.appliesToCohorts.isEmpty() }),
                specificInclusionCriteriaByTrialAndCohort,
                referencesByTrialAndId,
                database.unusedRulesToKeep.filter(validRules::contains).map(EligibilityRule::valueOf).toSet(),
                trialConfigDatabaseValidator.validate(database)
            )
        }

        private fun inclusionCriteriaConfigsByCohortId(
            inclusionCriteriaConfigs: List<InclusionCriteriaConfig>
        ): Map<String, List<InclusionCriteriaConfig>> {
            return inclusionCriteriaConfigs.flatMap { config -> config.appliesToCohorts.map { Pair(it, config) } }
                .groupBy({ it.first }, { it.second })
        }

        private fun <T : TrialConfig> configsByTrial(configs: List<T>): Map<String, List<T>> {
            return configs.groupBy(TrialConfig::trialId)
        }
    }
}